lib/common/socket/evloop.c.h

```diff
@@ -115,19 +115,56 @@ static void link_to_statechanged(struct st_h2o_evloop_socket_t *sock)
     }
 }
 
-static const char *on_read_core(int fd, h2o_buffer_t **input)
+static const char *on_read_core(struct st_h2o_evloop_socket_t *sock)
 {
     int read_any = 0;
+    int fd = sock->fd;
+    h2o_buffer_t **input = sock->super.ssl == NULL ? &sock->super.input : &sock->super.ssl->input.encrypted;
+    int timestamping = sock->super.timestamping;
+    long long timestamping_rate = sock->super.timestamping_rate;
+    long long *timestamping_rate_counter = sock->super.timestamping_rate_counter;
+
+    struct msghdr msg;
 
     while (1) {
         ssize_t rret;
         h2o_iovec_t buf = h2o_buffer_try_reserve(input, 4096);
+        size_t buflen = (buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 +1);
+
         if (buf.base == NULL) {
             /* memory allocation failed */
             return h2o_socket_error_out_of_memory;
         }
-        while ((rret = read(fd, buf.base, buf.len <= INT_MAX / 2 ? buf.len : INT_MAX / 2 + 1)) == -1 && errno == EINTR)
+        if (!timestamping) {
+          while ((rret = read(fd, buf.base, buflen)) == -1 && errno == EINTR)
```

Comment:
Is there a performance downside in using `recvmsg` all the time?

Always using `recvmsg`, while setting `msg_control` only when we use packet timestamping might make the code simpler.